features/192-asset-gear/index.md

```diff
@@ -0,0 +1,495 @@
+
+It supports both public and private asset.
+
+It can transform image asset on the fly via query string.
+
+The Asset Gear does not support static asset. Another gear is responsible for static asset.
+
+## Layout
+
+The Asset Gear uses a single bucket for all apps. The asset name is in the following format.
+
+```
+/<app-id>/<random-string>[.<ext>]
+```
+
+- `app-id`: The app ID.
+- `random-string`: A randomly generated string.
+- `ext`: The inferred file extension (if any) from the given content type.
+
+## HTTP API
+
+### POST /_asset/presign
+
+#### Description
+
+This endpoint returns a presigned upload request which can be used to upload an asset.
+
+This endpoint requires API Key and authenticated user.
+
+#### Request JSON Schema
+
+```json
+{
+  "type": "object",
+  "properties": {
+    "content_type": { "type": "string" },
+    "content_md5": { "type": "string" },
+    "content_length": { "type": "integer" },
+    "access": { "enum": ["public", "private"] }
+  },
+  "required": ["content_type", "content_md5", "content_length"]
+}
+```
+
+- `content_type`: The media type of the asset.
+- `content_md5`: The MD5 hash of the asset.
+- `content_length`: The size in bytes of the asset.
+- `access`: The access control of the asset. `public` is the default.
+
+##### Request Example
+
+```json
+{
+  "content_type": "image/png",
+  "content_md5": "Zum77CZrrrGRDM18nlplig==",
+  "content_length": 26419
+}
+```
+
+#### Response JSON Schema
+
+```json
+{
+  "type": "object",
+  "properties": {
+    "asset_id": { "type": "string" },
+    "url": { "type": "string" },
+    "method": { "type": "string" },
+    "headers": {
+      "type": "array",
+      "items": {
+        "type": "object",
+        "properties": {
+          "name": { "type": "string" },
+          "value": { "type": "string" }
+        }
+      }
+    }
+  },
+  "required": ["asset_id", "url", "method", "headers"]
+}
+```
+
+- `asset_id`: The asset ID. It becomes valid when the asset is uploaded.
+- `url`: The presigned URL to upload the asset.
+- `method`: The method to be used in the upload request.
+- `headers`: The headers to be included in the upload request.
+
+##### Response Example
+
+```json
+{
+  "asset_id": "myimage.png",
+  "url": "https://storage.skygearapps.com/bucket/myappid/myimage.png?...",
+  "method": "PUT",
+  "headers": [
+    { "name": "content-type", "value": "image/png" },
+    { "name": "content-md5", "value": "Zum77CZrrrGRDM18nlplig==" },
+    { "name": "content-length", "value": "26419" }
+  ]
+}
+```
+
+#### Specification
+
+1. Validate `content_type`.
+1. Let `name` be a randomly generated string.
+1. Let `ext` be the file extension derived from `content_type`.
+1. If `ext` is found, append it to `name`.
+1. Let `asset_id` be `/<app-id>/<name>`.
+1. Presign `asset_id` with `content_type`, `content_length`, `content_md5` as headers.
+1. Return the presigned request.
+
+### POST /_asset/sign
```

Comment:
It is not clear on the difference between `sign` and `presign`: both seems to sign and return an asset request. Do we need two API endpoint for this?

Reply:
`sign` is for signing an existing private asset so that it can be accessed via a signed URL.
`presign` is for generating a signed URL for upload. Carmen also have comment on this. Let me rename `presign` to clarify its purpose.